core/slice.go

```diff
@@ -978,3 +1028,196 @@ func (sl *Slice) AddPendingEtxsRollup(pEtxsRollup types.PendingEtxsRollup) error
+		rawdb.DeleteCanonicalHash(sl.sliceDb, header.NumberU64())
+		rawdb.DeleteHeaderNumber(sl.sliceDb, header.Hash())
+		rawdb.DeleteTermini(sl.sliceDb, header.Hash())
+		rawdb.DeleteEtxSet(sl.sliceDb, header.Hash(), header.NumberU64())
+		if nodeCtx != common.ZONE_CTX {
+			pendingEtxsRollup := rawdb.ReadPendingEtxsRollup(sl.sliceDb, header.Hash())
+			// First hash in the manifest is always a dom block and it needs to be
+			// deleted separately because last hash in the final iteration will be
+			// referenced in the next dom block after the restart
+			for _, manifestHash := range pendingEtxsRollup.Manifest[1:] {
+				rawdb.DeletePendingEtxs(sl.sliceDb, manifestHash)
+			}
+			rawdb.DeletePendingEtxs(sl.sliceDb, header.Hash())
+			rawdb.DeletePendingEtxsRollup(sl.sliceDb, header.Hash())
+		}
+		// delete the trie node for a given root of the header
+		rawdb.DeleteTrieNode(sl.sliceDb, header.Root())
+		badHashes = append(badHashes, header.Hash())
+		parent := sl.hc.GetHeader(header.ParentHash(), header.NumberU64()-1)
+		header = parent
+		if header.Hash() == hash || header.Hash() == sl.config.GenesisHash {
+			break
+		}
+	}
+
+	sl.AddToBadHashesList(badHashes)
+	// Set the current header
+	currentHeader = sl.hc.GetHeaderByHash(hash)
+	sl.hc.currentHeader.Store(currentHeader)
+
+	// Recover the snaps
+	if nodeCtx == common.ZONE_CTX {
+		sl.hc.bc.processor.snaps, _ = snapshot.New(sl.sliceDb, sl.hc.bc.processor.stateCache.TrieDB(), sl.hc.bc.processor.cacheConfig.SnapshotLimit, currentHeader.Root(), true, true)
+	}
+}
+
+func (sl *Slice) GenerateRecoveryPendingHeader(pendingHeader *types.Header, checkPointHashes []common.Hash) error {
+	nodeCtx := common.NodeLocation.Context()
+	if nodeCtx == common.PRIME_CTX {
+		for i := 0; i < common.NumRegionsInPrime; i++ {
+			if sl.subClients[i] != nil {
+				sl.subClients[i].GenerateRecoveryPendingHeader(context.Background(), pendingHeader, checkPointHashes)
+			}
+		}
+	} else if nodeCtx == common.REGION_CTX {
+		newPendingHeader := sl.SetHeadBackToRecoveryState(pendingHeader, checkPointHashes[common.NodeLocation.Region()])
+		for i := 0; i < common.NumZonesInRegion; i++ {
+			if sl.subClients[i] != nil {
+				sl.subClients[i].GenerateRecoveryPendingHeader(context.Background(), newPendingHeader.Header, newPendingHeader.Termini)
+			}
+		}
+	} else {
+		sl.SetHeadBackToRecoveryState(pendingHeader, checkPointHashes[common.NodeLocation.Zone()])
+	}
+	return nil
+}
+
+func (sl *Slice) ComputeRecoveryPendingHeader(hash common.Hash) types.PendingHeader {
+	block := sl.hc.GetBlockByHash(hash)
+	pendingHeader, err := sl.miner.worker.GeneratePendingHeader(block, false)
+	if err != nil {
+		log.Error("Error generating pending header during the checkpoint recovery process")
+		return types.PendingHeader{}
+	}
+	termini := sl.GenerateTermini(hash)
+	sl.bestPhKey = hash
+	return types.PendingHeader{Header: pendingHeader, Termini: termini}
+}
+
+func (sl *Slice) GenerateTermini(hash common.Hash) []common.Hash {
```

Comment:
Why does this function exist? If there is a good block we should be able to look up the termini.

Reply:
I wrote this function to generate bad hashes, testing with the Reading of Termini from the database, that should work